web/packages/teleterm/src/mainProcess/shell.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * Teleport
+ * Copyright (C) 2024 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import fs from 'node:fs';
+import os from 'node:os';
+
+import which from 'which';
+
+import Logger from 'teleterm/logger';
+
+export interface Shell {
+  /**
+   * Shell identifier, for example, pwsh.exe or zsh
+   * (it doesn't have to be the same as the binary name).
+   * Used as an identifier in the app config or in a document.
+   * Must be unique.
+   * */
+  id: string;
+  /** Friendly name, for example, Windows PowerShell, zsh. */
+  friendlyName: string;
+  /** Shell executable, for example, C:\\Windows\system32\pwsh.exe, /bin/zsh. */
+  binPath: string;
+}
+
+export async function getAvailableShells(): Promise<Shell[]> {
+  switch (process.platform) {
+    case 'linux':
+    case 'darwin':
+      return getUnixShells();
+    case 'win32': {
+      return getWindowsShells();
+    }
+  }
+}
+
+export function getDefaultShell(availableShells: Shell[]): string {
+  const logger = new Logger();
+  switch (process.platform) {
+    case 'linux':
+    case 'darwin': {
+      const fallbackShell = 'bash';
+      const { shell } = os.userInfo();
+      const shellId = availableShells.find(
+        availableShell => availableShell.binPath === shell
+      )?.id;
+
+      if (!shellId) {
+        logger.error(
+          `Failed to read ${process.platform} platform default shell, using fallback: ${fallbackShell}.\n`
+        );
+
+        return fallbackShell;
+      }
+
+      return shellId;
+    }
+    case 'win32':
+      if (availableShells.find(shell => shell.id === 'pwsh.exe')) {
+        return 'pwsh.exe';
+      }
+      return 'powershell.exe';
+  }
+}
+
+async function getUnixShells(): Promise<Shell[]> {
```

Comment:
I have zsh installed from homebrew, so on my Mac the list has a duplicate for zsh. On Ubuntu, `/etc/shells` looks as follows, so the list has duplicates too.

How could we address this? By showing full paths maybe? idk. But also why would we offer to the user a choice between `/bin/bash` and `/usr/bin/bash` or even let them pick rbash in the first place?

```
# /etc/shells: valid login shells
/bin/sh
/bin/bash
/usr/bin/bash
/bin/rbash
/usr/bin/rbash
/usr/bin/sh
/bin/dash
/usr/bin/dash
```

| macOS | Ubuntu |
| --- | --- |
| <img width="274" alt="shells-macos" src="https://github.com/user-attachments/assets/8480c3e6-84d5-4790-b03f-7ea17f96de7a"> | <img width="349" alt="shells-ubuntu" src="https://github.com/user-attachments/assets/257ed9b9-508c-4223-be63-7b2608e7a869"> |

Reply:
After our discussion, I replaced that with showing only the default shell on Unix + an option to select a custom one (using a file selector). I think it is much better now!

I also changed the context menu structure a bit, please let me know what do you think about it.

I haven't tested it yet on Linux and Windows, but the code can be reviewed.